core/src/main/java/org/apache/iceberg/avro/AvroSchemaUtil.java

```diff
@@ -196,15 +199,38 @@ static Schema createProjectionMap(String recordName,
     return LogicalMap.get().addToSchema(Schema.createArray(keyValueRecord));
   }
 
-  private static int getId(Schema schema, String propertyName) {
+
+  static Integer getId(Schema schema, String propertyName) {
+    Integer id = getId(schema, propertyName, null, null);
+    Preconditions.checkNotNull(id, "Missing expected '%s' property", propertyName);
+    return id;
+  }
+
+  static Integer getId(Schema schema, String propertyName,
+                   NameMapping nameMapping, List<String> names) {
     if (schema.getType() == UNION) {
-      return getId(fromOption(schema), propertyName);
+      return getId(fromOption(schema), propertyName, nameMapping, names);
     }
 
     Object id = schema.getObjectProp(propertyName);
-    Preconditions.checkNotNull(id, "Missing expected '%s' property", propertyName);
+    if (id != null) {
+      return toInt(id);
+    } else {
+      Preconditions.checkArgument(nameMapping != null,
+          "Schema does not have field id and name mapping not specified");
```

Comment:
I don't think this should thrown an exception. If the name mapping isn't present or doesn't have the given field, then this should return null so that the field is pruned.

I think it should be like this:

```java
  private static Integer getId(Schema schema, String propertyName, NameMapping nameMapping, List<String> names) {
    if (schema.getType() == UNION) {
      return getId(fromOption(schema), propertyName, nameMapping, names);
    }

    Object id = schema.getObjectProp(propertyName);
    if (id != null) {
      return toInt(id);
    } else if (nameMapping != null) {
      MappedField mappedField = nameMapping.find(names);
      if (mappedField != null) {
        return mappedField.id();
      }
    }

    return null;
  }
```